Capstone/Backend/AssessmentPlatform/src/main/java/com/nucleusteq/asessmentPlatform/services/impl/CategoryServiceImpl.java

```diff
@@ -0,0 +1,160 @@
+    @Autowired
+    private ModelMapper modelMapper;
+    private Logger logger = LoggerFactory.getLogger(CategoryController.class);
+    /**
+     * Adds a new category.
+     *
+     * @param categoryDto The CategoryDto containing category details to be
+     *                    added.
+     * @return The added CategoryDto.
+     * @throws DuplicateResourceException If a category with the same title
+     *                                    already exists.
+     */
+    @Override
+    public final CategoryDto addCategory(final CategoryDto categoryDto) {
+        Category category = this.dtoToCategory(categoryDto);
+        Optional<Category> existingCategory = categoryRepo
+                .findByTitle(category.getTitle());
+        if (existingCategory.isPresent()) {
+            logger.error("Category with Title already exist");
+            throw new DuplicateResourceException("Category with title '"
+                    + category.getTitle() + "' already exists.");
+        }
+        categoryRepo.save(category);
+        logger.info("Category saved in Category Repository.");
+        return this.categoryToDto(category);
+    }
+
+    /**
+     * Retrieves all categories.
+     *
+     * @return A list of CategoryDtos containing all categories.
+     */
+    @Override
+    public final List<CategoryDto> getAllCategories() {
+        List<Category> categories = this.categoryRepo.findAll();
+        List<CategoryDto> categoryDtos = categories.stream()
+                .map(category -> this.categoryToDto(category))
+                .collect(Collectors.toList());
+        logger.info("Get All Category");
+        return categoryDtos;
+    }
+
+    /**
+     * Retrieves a category by its ID.
+     *
+     * @param id The ID of the category to retrieve.
+     * @return The CategoryDto representing the retrieved category.
+     * @throws ResourceNotFoundException If the category with the specified ID
+     *                                   is not found.
+     */
+    @Override
+    public final CategoryDto getCategoryById(final int id) {
+        Category category = categoryRepo.findById(id)
+                .orElseThrow(() -> new ResourceNotFoundException(
+                        "Category not found with id " + id));
+        logger.info("Get Category by Category Id");
+        return this.categoryToDto(category);
+    }
+
+    /**
+     * Updates a category with the given ID.
+     *
+     * @param category The updated CategoryDto.
+     * @param id       The ID of the category to update.
+     * @return The updated CategoryDto.
+     * @throws ResourceNotFoundException If the category with the specified ID
+     *                                   is not found.
+     */
+
+    @Override
+    public final CategoryDto updateCategory(final CategoryDto category,
+            final int id) {
+        Category updatedCategory = categoryRepo.findById(id).orElseThrow(
+                () -> new ResourceNotFoundException("category not found"));
```

Comment:
if there is a constant message don't pass it directly , store it in a class and the call the class